3scale/3scale-porta-go-client
# [WIP] Addition of delete functinailty for app.go

## What 
Addition of delete functionality which is required for [THREESCALE-7668](https://issues.redhat.com/browse/THREESCALE-7668)

## Verification 
1. Checkout out this branch.
2.  Create an Application in 3scale on RHOAM cluster
3. Retrieve ID of Application and ID of account under which the application was created  ( use Application List  API call in 3scale API docs)
4. Create a new test project and run the following command to generate go.mod
```bash
go mod init github.com/Patryk-Stefanski/porta-test
go get github.com/3scale/3scale-porta-go-client/client@v0.6.0
```
5. Replace the imported porta client with the local version that you checked out,
```bash
 go mod edit -replace="github.com/3scale/3scale-porta-go-client@v0.6.0=<Path_To_Local_Porta_Client>
```
6.  go mod should look something like this 
```bash
module github.com/Patryk-Stefanski/porta-test

go 1.13

require github.com/3scale/3scale-porta-go-client v0.6.0

replace github.com/3scale/3scale-porta-go-client v0.6.0 => /home/PStefans/Git/Github/3scale/3scale-porta-go-client
```

7. Create a main.go file with the following 
** Note : Some functions might need to be changed to accound for different account , plan IDs etc .... **
```bash
package main

import (
	"encoding/json"
	"fmt"
	"github.com/3scale/3scale-porta-go-client/client"
)

func jsonPrint(obj interface{}) {
	jsonData, err := json.MarshalIndent(obj, "", "  ")
	if err != nil {
		panic(err)
	}
	fmt.Println(string(jsonData))
}

func main() {
	testDelete()
}

func testDelete() {
	adminPortalURL := "https://3scale-admin.apps.pstefans2.***"
	threescaleAccessToken := "*************"

	adminPortal, err := client.NewAdminPortalFromStr(adminPortalURL)
	if err != nil {
		fmt.Printf("%+v\n", err)
		return
	}
	threescaleClient := client.NewThreeScale(adminPortal, threescaleAccessToken, nil)

	err = threescaleClient.DeleteApplication(3, 9) // err should not be nil when the product exists
	if err != nil {
		fmt.Printf("%+v\n", err)
		return
	}
}

func testUpdate() {
	adminPortalURL := "https://3scale-admin.apps.pstefans2.y9c8.s1.devshift.org"
	threescaleAccessToken := "DwRuaEZXyPMK9iSG"

	var params = client.Params{"description": "example"}

	adminPortal, err := client.NewAdminPortalFromStr(adminPortalURL)
	if err != nil {
		fmt.Printf("%+v\n", err)
		return
	}
	threescaleClient := client.NewThreeScale(adminPortal, threescaleAccessToken, nil)

	_, err = threescaleClient.UpdateApplication(3, 9, params) // err should not be nil when the product exists
	if err != nil {
		fmt.Printf("%+v\n", err)
		return
	}
}

func testChangeApplicationPlan() {
	adminPortalURL := "https://3scale-admin.apps.pstefans2.y9c8.s1.devshift.org"
	threescaleAccessToken := "DwRuaEZXyPMK9iSG"

	adminPortal, err := client.NewAdminPortalFromStr(adminPortalURL)
	if err != nil {
		fmt.Printf("%+v\n", err)
		return
	}
	threescaleClient := client.NewThreeScale(adminPortal, threescaleAccessToken, nil)

	_, err = threescaleClient.ChangeApplicationPlan(3, 9, 7) // err should not be nil when the product exists
	if err != nil {
		fmt.Printf("%+v\n", err)
		return
	}
}

func testCreateApplicationCustomPlan() {
	adminPortalURL := "https://3scale-admin.apps.pstefans2.y9c8.s1.devshift.org"
	threescaleAccessToken := "DwRuaEZXyPMK9iSG"

	adminPortal, err := client.NewAdminPortalFromStr(adminPortalURL)
	if err != nil {
		fmt.Printf("%+v\n", err)
		return
	}
	threescaleClient := client.NewThreeScale(adminPortal, threescaleAccessToken, nil)

	_, err = threescaleClient.CreateApplicationCustomPlan(3, 9) // err should not be nil when the product exists
	if err != nil {
		fmt.Printf("%+v\n", err)
		return
	}
}

func testDeleteApplicationCustomPlan() {
	adminPortalURL := "https://3scale-admin.apps.pstefans2.y9c8.s1.devshift.org"
	threescaleAccessToken := "DwRuaEZXyPMK9iSG"

	adminPortal, err := client.NewAdminPortalFromStr(adminPortalURL)
	if err != nil {
		fmt.Printf("%+v\n", err)
		return
	}
	threescaleClient := client.NewThreeScale(adminPortal, threescaleAccessToken, nil)

	err = threescaleClient.DeleteApplicationCustomPlan(3, 9) // err should not be nil when the product exists
	if err != nil {
		fmt.Printf("%+v\n", err)
		return
	}
}

func testApplicationSuspend() {
	adminPortalURL := "https://3scale-admin.apps.pstefans2.y9c8.s1.devshift.org"
	threescaleAccessToken := "DwRuaEZXyPMK9iSG"

	adminPortal, err := client.NewAdminPortalFromStr(adminPortalURL)
	if err != nil {
		fmt.Printf("%+v\n", err)
		return
	}
	threescaleClient := client.NewThreeScale(adminPortal, threescaleAccessToken, nil)

	_, err = threescaleClient.ApplicationSuspend(3, 9) // err should not be nil when the product exists
	if err != nil {
		fmt.Printf("%+v\n", err)
		return
	}
}

func testApplicationResume() {
	adminPortalURL := "https://3scale-admin.apps.pstefans2.y9c8.s1.devshift.org"
	threescaleAccessToken := "DwRuaEZXyPMK9iSG"

	adminPortal, err := client.NewAdminPortalFromStr(adminPortalURL)
	if err != nil {
		fmt.Printf("%+v\n", err)
		return
	}
	threescaleClient := client.NewThreeScale(adminPortal, threescaleAccessToken, nil)

	_, err = threescaleClient.ApplicationResume(3, 9) // err should not be nil when the product exists
	if err != nil {
		fmt.Printf("%+v\n", err)
		return
	}
}

```

9.  Run main.go 
```bash
go run main.go
```

10.  Check that each of the functionalities works as expected by changing the function being called in the main function and checking that the function was completed in 3scale ui.
